akka-persistence-testkit/src/main/scala/akka/persistence/testkit/scaladsl/ImMemStorage.scala

```diff
@@ -0,0 +1,350 @@
+    currentPolicy.tryProcess(persistenceId, ReadSeqNum) match {
+      case ProcessingSuccess  ⇒ reloadHighestSequenceNum(persistenceId)
+      case Reject(ex)         ⇒ throw ex
+      case StorageFailure(ex) ⇒ throw ex
+    }
+  }
+
+  def tryDelete(persistenceId: String, toSeqNumber: Long): Unit = {
+    currentPolicy.tryProcess(persistenceId, Delete(toSeqNumber)) match {
+      case ProcessingSuccess  ⇒ deleteToSeqNumber(persistenceId, toSeqNumber)
+      case Reject(ex)         ⇒ throw ex
+      case StorageFailure(ex) ⇒ throw ex
+    }
+  }
+
+}
+
+object InMemStorageEmulator {
+
+  trait JournalOperation
+
+  case class Read(batch: immutable.Seq[Any]) extends JournalOperation
+
+  case class Write(batch: immutable.Seq[Any]) extends JournalOperation
+
+  case object ReadSeqNum extends JournalOperation
+
+  case class Delete(toSeqNumber: Long) extends JournalOperation
+
+  object JournalPolicies extends DefaultPolicies[JournalOperation]
+
+}
+
+trait SnapShotStorageEmulator extends TestKitStorage[(SnapshotMetadata, Any), SnapshotOperation] {
+  import SnapShotStorageEmulator._
+
+  override def mapAny(key: String, elems: immutable.Seq[Any]): immutable.Seq[(SnapshotMetadata, Any)] = {
+    val sn = reloadHighestSequenceNum(key)
+    elems.zipWithIndex.map(p ⇒ (SnapshotMetadata(key, p._2 + sn), p._1))
+  }
+
+  override def reprToSeqNum(repr: (SnapshotMetadata, Any)): Long = repr._1.sequenceNr
+
+  override protected val DefaultPolicy = SnapshotPolicies.PassAll
+
+  def tryAdd(meta: SnapshotMetadata, payload: Any): Unit = {
+    currentPolicy.tryProcess(meta.persistenceId, Write(payload)) match {
+      case ProcessingSuccess ⇒
+        add(meta.persistenceId, (meta, payload))
+        Success(())
+      case f: ProcessingFailure ⇒ throw f.error
+
+    }
+  }
+
+  def tryRead(persistenceId: String, criteria: SnapshotSelectionCriteria): Option[SelectedSnapshot] = {
+    val selectedSnapshot =
+      read(persistenceId)
+        .flatMap(_.reverseIterator.find(v ⇒ criteria.matches(v._1))
+          .map(v ⇒ SelectedSnapshot(v._1, v._2)))
+    currentPolicy.tryProcess(persistenceId, Read(selectedSnapshot.map(_.snapshot))) match {
+      case ProcessingSuccess    ⇒ selectedSnapshot
+      case f: ProcessingFailure ⇒ throw f.error
+    }
+  }
+
+  def tryDelete(persistenceId: String, selectionCriteria: SnapshotSelectionCriteria): Unit = {
+    currentPolicy.tryProcess(persistenceId, DeleteByCriteria(selectionCriteria)) match {
+      case ProcessingSuccess    ⇒ delete(persistenceId, v ⇒ selectionCriteria.matches(v._1))
+      case f: ProcessingFailure ⇒ throw f.error
+    }
+  }
+
+  def tryDelete(meta: SnapshotMetadata): Unit = {
+    currentPolicy.tryProcess(meta.persistenceId, DeleteSnapshot(meta)) match {
+      case ProcessingSuccess    ⇒ delete(meta.persistenceId, _._1.sequenceNr == meta.sequenceNr)
+      case f: ProcessingFailure ⇒ throw f.error
+    }
+  }
+
+}
+
+object SnapShotStorageEmulator {
```

Comment:
I think we spell it `Snapshot` in other places

Reply:
Got it!